(chore) Enhance ClockUtil with clock functionality

stefanzilske opened the pull request:
I folks, 

I made a few changes to ClockUtil. 

Short story: Instead of freezing it to a fixed time with setCurrentTime, it now contains a Clock, which is shifted by an offset. Advantage: After time travelling, the time moves on, which is especially useful when testing jobs (use case: travel to the future, schedule a job and it will magically be executed, because time is not frozen anymore).

The old API still exists and is delegated to the new clock.

Let me know what you think about this pull request, feedback is welcome!

stefanzilske commented:
I adapted the PR once more, to retain the old freeze behaviour with `setCurrentTime()`, so all tests, that rely on a frozen time will still work. For the described use case the new `offset()` can be used.

yanavasileva commented:
Hi Stefan,

We discussed this PR internally and as we getting closer to building the next minor release,  we decided to postpone it for 7.11.
As soon as we start working on the new topics for 7.11, we will reconsider this one as well.

Best regards,
Yana

yanavasileva commented:
Hi Stefan,

Happy New Year, I wish you all the best!

As I promised I am going to merge your change.
I have some feedback:
* I think `sameSecondOfMinute` assertions should be replaced with assertion check of the whole date including seconds. In my opinion it is not enough to check only the seconds, if a problem with the day/hour of date exists it won't be catch with these tests.
* Also could you please add the following license header to the newly created ClockUtilTest class:

```java
/*
 * Copyright © 2019 camunda services GmbH and various authors (info@camunda.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

Thanks again and Best regards,
Yana

stefanzilske commented:
Hi @yanavasileva 

a happy new year to you as well!

I will update the assertions. Since the clock is now moving forward after shifting the offset, an equals-assertions would fail, but I agree that asserting only the second is not sufficient. AssertJ by the way has much more advanced time assertions capabilities than Hamcrest, but I will do my best to achieve the same 😉

I will also add the license header and update the pull request when done.

Cheers,
Stefan

stefanzilske commented:
@yanavasileva please have a look at my latest commit

yanavasileva commented:
Hi Stefan,

Unfortunately 2.* versions of Hamcrest (where `#same*()` methods are introduced) require Java 8, I forgot to mention it in my previous comment.
Do you want change it with something else or I can adjust it by myself?

Best regards,
Yana

stefanzilske commented:
Hi,

I replaced the Hamcrest assertions (and removed hamcres-date dependency) with AssertJ, which I found was already in parent-pom in v1.5.0. I hope there are no more clashes with Java version anymore.

Btw. when will Camunda eventually be Java8 compatible?

Cheers,
Stefan

yanavasileva commented:
Related to https://app.camunda.com/jira/browse/CAM-9624

yanavasileva commented:
Hi Stefan,

Many thanks for the adjustments, I will merge it this week.
Regarding Java 8 compatibility - for now it is not planned.

Best regards,
Yana

yanavasileva commented:
Hi Stefan,

I merged the changes, I needed to adjust formatting, code style according to Camunda specifics.
Hereby I am closing the PR, in case of any questions or thoughts, please do not hesitate to reopen it.
Thanks once again for the contribution.

Best regards,
Yana
